### PR TITLE
Azure Service Fabric based new provider implementation.

### PR DESCRIPTION
**Pull request by kamesh-msft**
This pull request will introduce new provider based on Azure Service Fabric. 

* Scheduled Tasks, Timers, Sub orchestrations work!
* Fabric based instance store has limited functionality for now!
  * No support for history events.
  * Only the latest state for a given Orchestration instance is stored (previous state changes are overwritten by latest). [For a period of 24 hour time window]
  * State can be queried only for Orchestrations still running / pending OR completed within an hour ago.

For more detailed notes refer to src\DurableTask.AzureServiceFabric\ReadMe.md

### CODE REVIEW

File: src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
+    <PackageReference Include="ImpromptuInterface" Version="6.2.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="7.0.1" />
+    <Reference Include="System.Web" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
+    <!-- Restoring packages for netstandard2.0 causes warnings. As warnigns are treated as errors, compilation will fail. -->
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj]: >warnigns [](start = 67, length = 8)

nit: typo

File: src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    <!-- Once the packages support netstandard2.0, this project will support netstandard2.0. -->
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj]: >the packages [](start = 14, length = 12)

which packages?

File: src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        {
+            return GetInstanceStoreBackupDictionaryKey(time, TimeFormatStringPrefix);
+        }
+
+        string GetInstanceStoreBackupDictionaryKey(DateTime time, string formatString)
+        {
+            return InstanceStoreCollectionNamePrefix + time.ToString(formatString);
+        }
+
+        Task CleanupDayOldDictionaries()
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs]: >CleanupDayOldDictionaries [](start = 13, length = 25)

Async

File: src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+                var purgeTime = GetDictionaryKeyFromTimePrefixFormat(DateTime.UtcNow - TimeSpan.FromDays(1));
+
+                for (int i = 0; i < 24; i++)
+                {
+                    await this.stateManager.RemoveAsync($"{purgeTime}{i:D2}");
+                }
+            }, initialDelay: TimeSpan.FromMinutes(5), delayOnSuccess: TimeSpan.FromHours(12), delayOnException: TimeSpan.FromHours(1), actionName: $"{nameof(CleanupDayOldDictionaries)}", token: this.cancellationToken);
+        }
+
+        Task CleanupOldDictionaries()
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs]: >CleanupOldDictionaries [](start = 13, length = 22)

Async

File: src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+                    var deleteTime = await Utils.MeasureAsync(async () =>
+                    {
+                        await this.stateManager.RemoveAsync(storeName);
+                    });
+                    ServiceFabricProviderEventSource.Tracing.LogTimeTaken($"Deleting reliable state {storeName}", deleteTime.TotalMilliseconds);
+                }
+            }, initialDelay: TimeSpan.FromMinutes(5), delayOnSuccess: TimeSpan.FromHours(1), delayOnException: TimeSpan.FromMinutes(10), actionName: $"{nameof(CleanupOldDictionaries)}", token: this.cancellationToken);
+        }
+
+        async Task EnsureStoreInitialized()
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/FabricOrchestrationInstanceStore.cs]: >EnsureStoreInitialized [](start = 19, length = 22)

Async

File: src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj (hunk 2 above)
Comment (kamesh-msft) [reply src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj]: image

Comment (kamesh-msft) [reply src/DurableTask.AzureServiceFabric/DurableTask.AzureServiceFabric.csproj]: This happens only in RELEASE mode.

File: src/DurableTask.AzureServiceFabric/Service/ProxyServiceExceptionsLogger.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Web.Http.ExceptionHandling;
+
+    using DurableTask.AzureServiceFabric.Tracing;
+
+    /// <summary>
+    /// Traces application exceptions.
+    /// </summary>
+    public class ProxyServiceExceptionsLogger : ExceptionLogger
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/Service/ProxyServiceExceptionsLogger.cs]: >Exceptions [](start = 29, length = 10)

Here and a few other places, use singular in classnames not plural. Also SessionMessagesProvider, SessionsProvider.

File: src/DurableTask.AzureServiceFabric/Service/DefaultStringPartitionHashing.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Implements <see cref="IPartitionHashing{T}"/> for string elements.
+    /// </summary>
+    public class DefaultStringPartitionHashing : IPartitionHashing<string>
+    {
+        /// <inheritdoc/>
+        public Task<long> GeneratePartitionHashCode(string value, CancellationToken cancellationToken)
```
Comment (simonporter) [inline src/DurableTask.AzureServiceFabric/Service/DefaultStringPartitionHashing.cs]: >Task [](start = 15, length = 4)

Interested why a hashing func would return a task. If this should be like this naming should have async suffix

**Pull request merged**